keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2150 @@
+
+A persistent ResourceClaim has a lifecyle that is independent of any particular
+pod. It gets created and deleted by the user. This is useful for resources
+which are expensive to configure and that can be used multiple times by pods,
+either at the same time or one after the other. Such persistent ResourceClaims
+get referenced in the pod spec by name. When a PodTemplateSpec in an app
+controller spec references a ResourceClaim by name, all pods created by that
+controller also use that name and thus share the resources allocated for that
+ResourceClaim.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.ResourceClaims
+list can also be a ResourceClaimTemplate. When a Pod gets created, such a
+template will be used to create a normal ResourceClaim with the Pod as owner
+with an
+[OwnerReference](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference)),
+and then the normal allocation of the resource takes place. Once the pod got
+deleted, the Kubernetes garbage collector will also delete the
+ResourceClaim.
+
+The difference between persistent and ephemeral resources for kube-scheduler
+and kubelet is that the name of the ResourceClaim needs to be determined
+differently: the name of an ephemeral ResourceClaim is `<pod>-<resource name in
+the pod spec>`. Ownership must be checked to detect accidental conflicts with
+persistent ResourceClaims or previous incarnations of the same ephemeral
+resource. This is the same approach that was chosen for [generic ephemeral
+volumes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md).
+For a resource driver there is no difference.
+
+### Coordinating resource allocation through the scheduler
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. The downside is that pod scheduling is less flexible.
+
+For delayed allocation, a node is selected tentatively by the scheduler
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource
+            is allocated and unused,
```

Comment:
what is unused in terms of fields?

Reply:
"allocated" and "in use" occur so often in the text that always referencing the fields would have become distracting (IMHO). I instead added a definition to the "State and communication" section:

The entire state of a resource can be determined by looking at its
ResourceClaim (see [API below](#api) for details), for example:

- It is allocated if and only if `ResourceClaimStatus.Allocated` is non-nil and
  points to the `AllocationResult`.

- It is in use if and only if `ResourceClaimStatus.ReservedFor` contains one or
  more users.  It does not matter whether those users, usually pods, are
  currently running because that could change at any time.

- A resource is no longer needed when it has a `DeletionTimestamp`. It must not
  be freed yet when it is still in use.